### DISCUSSION TITLE
Marble Solitary

### DISCUSSION THREAD

**Discussion by DevBefell** [beeware/toga#1258] [Show and tell]
I created Marble/Peg Solitary Game in Toga, with buttons and colors, relying on ids for placements:

https://user-images.githubusercontent.com/54691311/117893551-793bac80-b26f-11eb-8bf9-d424eb587c96.mp4

I didn't do any comments so take your time to read the code, and suggestions are welcomed:
https://gist.github.com/PyICoder/f0a77b5177bfbdefc6ebae089829ef35

**Comment by freakboy3742** [comment]
Neat - Thanks for sharing!

The only comment I'd have on the code is that the list of buttons could probably be procedurally generated (i.e., a pair of for loops creating buttons, rather than 36 explicit creations statements) - but that's an optimization, not anything critical. 

Another (and more Toga-specific) cleanup would be to combine the Button creation call with the `add` call; i.e.:
```
box0 = toga.Box(children=[
    toga.Button(label="", style=Pack(width=32, height=32), on_press=self.on_button_press, id="0|2"),
    toga.Button(label="", style=Pack(width=32, height=32), on_press=self.on_button_press, id="0|3"),
    toga.Button(label="", style=Pack(width=32, height=32), on_press=self.on_button_press, id="0|4"),
])
```
The only reason to explicitly create and assign the widgets to variables is if you need to reference the widget later. However, in your case, you're handling everything with the `on_button_press` handler, so there's no need to keep those references.

My question to you: any feedback on the Toga? Any features or changes that you'd like to see that would have made the development experience better?

**Reply by DevBefell** [reply]
Ah I see, didn't know you can do that, thanks. 
My answer for your question is, I would want to see events handlers similar to Tkinter's event handlers, but overall Toga seems very cool, although I can tell it is still in development, but very promising.

**Reply by freakboy3742** [reply]
Can you give an example of what you mean by that? Tkinter's event handlers are effectively callables with specific arguments (the widget generating the event, and some event data) - which is very similar to what Toga generates (that's what `on_press` is doing). What feature in particular are you looking for? Or have I misunderstood what you're referring to?

**Reply by DevBefell** [reply]
For example, something binded to key press:
```py
import tkinter

def something(event):
    pass

root = tkinter.Tk()
root.bind("<KeyPress>", something)
root.mainloop()
```
Edit: forgot the event param
